conf/closed_loop.bess

```diff
@@ -0,0 +1,501 @@
+    m_green  = 1 # For green traffic
+    m_yellow = 2 # For yellow traffic
+    m_red    = 3 # For red traffic
+    # User defined gates
+    m_fail   = 4 # For lookup failure traffic
+    m_unmeter = 5 # For unmetered traffic
+    # Admit green, yellow, unmetered and lookup failures, drop red.
+    sliceMeter:m_green -> executeFAR
+    sliceMeter:m_yellow -> executeFAR
+    sliceMeter:m_red -> sliceMeterRed::Sink()
+    sliceMeter:m_fail -> executeFAR
+    sliceMeter:m_unmeter -> executeFAR
+    sliceMeter.set_default_gate(gate=m_fail)
+    _in = sliceMeter
+
+farLookup::ExactMatch(fields=[{'attr_name':'far_id', 'num_bytes':4}, \
+                              {'attr_name':'fseid', 'num_bytes':8}], \
+                      values=[{'attr_name':'action', 'num_bytes':1}, \
+                              {'attr_name':'tunnel_out_type', 'num_bytes':1}, \
+                              {'attr_name':'tunnel_out_src_ip4addr', 'num_bytes':4}, \
+                              {'attr_name':'tunnel_out_dst_ip4addr', 'num_bytes':4}, \
+                              {'attr_name':'tunnel_out_teid', 'num_bytes':4}, \
+                              {'attr_name':'tunnel_out_udp_port', 'num_bytes':2}],\
+                      entries=parser.table_size_far_lookup):noGTPUEncap \
+    -> farMerge::Merge() \
+    -> _in
+
+# sessionQERLookup enforces a per UE, per direction meter rate limit
+sessionQERLookup::Qos(fields=[{'attr_name':'src_iface', 'num_bytes':1}, \
+                              {'attr_name':'fseid', 'num_bytes':8}],\
+                      entries=parser.table_size_session_qer_lookup)
+# Admit green, yellow and misses, drop red
+sessionQERLookup:qerGreenGate -> farLookup
+sessionQERLookup:qerYellowGate -> farLookup
+sessionQERLookup:qerRedGate -> sessionQERMeterRed::Sink()
+sessionQERLookup:qerStatusDropGate -> sessionQERStatusDrop::Sink()
+sessionQERLookup:qerUnmeteredGate -> farLookup
+sessionQERLookup:qerFailGate -> farLookup
+sessionQERLookup.set_default_gate(gate=qerFailGate)
+
+# Add logical pipeline when gtpuencap is needed
+farLookup:GTPUEncap \
+    -> gtpuEncap::GtpuEncap(add_psc=parser.gtppsc):1 \
+    -> outerL4Cksum::L4Checksum() \
+    -> outerIPCksum::IPChecksum() \
+    -> farMerge
+
+notify = UnixSocketPort(name='notifyCP', path=parser.notify_sockaddr)
+pfcpPort = UnixSocketPort(name='pfcpPort', path=parser.endmarker_sockaddr)
+pfcpPI::PortInc(port='pfcpPort') -> pfcpPI_timestamp::Timestamp() -> ports[parser.access_ifname].rtr
+executeFAR:farNotifyCPAction -> pfcpDetails::GenericEncap(fields=[ {'size': 8, 'attribute': 'fseid'}]) \
+                             -> farNotifyCP::PortOut(port='notifyCP')
+# Drop unknown packets
+pktParse:0 -> badPkts::Sink()
+appQERLookup:qerGreenGate -> sessionQERLookup
+appQERLookup:qerYellowGate -> sessionQERLookup
+appQERLookup:qerRedGate -> appQERMeterRed::Sink()
+appQERLookup:qerStatusDropGate -> packetReplica::Replicate(gates=[0,1])
```

Comment:
The change seems to be a bit localized, could we parametrize it in the `up4.bess` file instead with a default to disabled? It would be easier to maintain going forward if we change the pipeline, we would not need to change it in 2 files.

Reply:
I thought on doing what you suggest but this is a "customized" UPF pipeline (not generic), and this is why I wanted to keep it separated from the actual UPF pipeline (up4). That is, depending on the use case, weather to use one pipeline (up4) or the other

Reply:
Ok, in that case I do not have a problem with this.